client.go

```diff
@@ -28,19 +30,29 @@ type Client struct {
 	mutex       sync.Mutex
 }
 
+type RemoteAddrCloser interface {
+	RemoteAddr() net.Addr
+	Close() error
+}
+
 // NewClient creates a new client info container
-func NewClient(addr string) *Client {
+func NewClient(conn RemoteAddrCloser) *Client {
 	now := time.Now()
 	return &Client{
-		ID:          atomic.AddUint64(&clientInc, 1),
-		RemoteAddr:  addr,
 		firstAccess: now,
 		lastAccess:  now,
+		ID:          atomic.AddUint64(&clientInc, 1),
+		RemoteAddr:  conn.RemoteAddr().String(),
+		conn:        conn,
 	}
 }
 
-// Close will disconnect the client when the buffer has been send
-func (i *Client) Close() { i.closed = true }
+// Close will close client
+func (i *Client) Close() {
+	i.mutex.Lock()
```

Comment:
There maybe concurrently read/written for i.conn when server close client forcely.